src/ophyd_async/epics/pmac/_utils.py

```diff
@@ -40,63 +178,112 @@ def from_slice(cls, slice: Slice[Motor], ramp_up_time: float) -> _Trajectory:
+        durations: npt.NDArray[np.float64] = np.empty(trajectory_size, float)
+        # Default to program 1 to assume we acquire at all points
+        user_programs: npt.NDArray[np.int32] = np.empty(trajectory_size, float)
 
         # Ramp up time for start of collection window
         durations[0] = int(ramp_up_time / TICK_S)
-        # Half the time per point
-        durations[1:] = np.repeat(slice_duration / (2 * TICK_S), 2)
 
-        # Fill profile assuming no gaps
-        # Excluding starting points, we begin at our next frame
-        half_durations = slice_duration / 2
+        # Fill trajectory
         for motor in motors:
-            # Set the first position to be lower bound, then
-            # alternate mid and upper as the upper and lower
-            # bounds of neighbouring points are the same as gap is false
-            positions[motor][0] = slice.lower[motor][0]
-            positions[motor][1::2] = slice.midpoints[motor]
-            positions[motor][2::2] = slice.upper[motor]
-            # For velocities we will need the relative distances
-            mid_to_upper_velocities = (
-                slice.upper[motor] - slice.midpoints[motor]
-            ) / half_durations
-            lower_to_mid_velocities = (
-                slice.midpoints[motor] - slice.lower[motor]
-            ) / half_durations
-            # First velocity is the lower -> mid of first point
-            velocities[motor][0] = lower_to_mid_velocities[0]
-            # For the midpoints, we take the average of the
-            # lower -> mid and mid-> upper velocities of the same point
-            velocities[motor][1::2] = (
-                lower_to_mid_velocities + mid_to_upper_velocities
-            ) / 2
-            # For the upper points, we need to take the average of the
-            # mid -> upper velocity of the previous point and
-            # lower -> mid velocity of the current point
-            velocities[motor][2:-1:2] = (
-                mid_to_upper_velocities[:-1] + lower_to_mid_velocities[1:]
-            ) / 2
-            # For the last velocity take the mid to upper velocity
-            velocities[motor][-1] = mid_to_upper_velocities[-1]
+            # Keeps track of where we are in the trajectory
+            index_into_trajectory = 0
+            # Iterate over collection windows or gaps
+            for segment in segments:
+                # This inserts slice points or gap points into the output arrays
+                # Returns a new index into our trajectory arrays
+                # where more points can be inserted
+                index_into_trajectory = segment.insert_points_into_trajectory(
+                    index_into_trajectory=index_into_trajectory,
+                    positions=positions,
+                    velocities=velocities,
+                    durations=durations,
+                    user_programs=user_programs,
+                    motor=motor,
+                )
```

Comment:
We know the length of a GapSegment (`gap_length`) and CollectionWindow (`end - start`), so maybe we just implement `__len__` on them both and do

```suggestion
                segment.insert_points_into_trajectory(
                    index=index_into_trajectory,
                    positions=positions,
                    velocities=velocities,
                    durations=durations,
                    user_programs=user_programs,
                    motor=motor,
                )
                
                index_into_trajectory += len(segment)
```

rather than returning the updated index? Does that work or am I missing something?